lua/nvim-treesitter/ts_utils.lua

```diff
@@ -360,25 +360,28 @@ function M.swap_nodes(node_or_range1, node_or_range2, bufnr, cursor_to_second)
       range1["end"].line < range2.start.line
       or (range1["end"].line == range2.start.line and range1["end"].character < range2.start.character)
     then
-      line_delta = #text2 - #text1
+      line_delta = (range2["end"].line - range2["start"].line + 1) - (range1["end"].line - range1["start"].line + 1)
     end
 
     if range1["end"].line == range2.start.line and range1["end"].character < range2.start.character then
+      local lines1 = vim.split(text1, "\n")
+      local lines2 = vim.split(text2, "\n")
```

Comment:
this isn't the fastest, could try to figure out something faster (like counting the characters from the end of the string to a `\n`), but this is very much of an edge case and I don't expect people to be swapping very large nodes, so the extra memory usage should be ok? (famous last words)

Reply:
Pattern matching would also work, something like:

```lua
local function length_last_line(str)
  if str == "" then
    return 0
  end
  local first, last = string.find(str, "\n?[^\n]+$")
  return last - first
end

-- ...

local length_last_line1 = length_last_line(text1)
local length_last_line2 = length_last_line(text2)
```

Though that isn't super readable 🤷

Reply:
Wasn't aware that get_node_text wasn't no longer working on lines. Seems a bit odd for me to join the lines just to split them afterwards which occurs in many occasions of get_node_text. Maybe we should just revert it. Can't you get the #lines from the node range?

Reply:
You should get the length of the last line by looking at the start/end cols of the node range

Reply:
> You should get the length of the last line by looking at the start/end cols of the node range

OMG that's right. Sorry I'll fix that later today.

Reply:
Fixed it. I feel like it'd be cool to have some tests for this, but I assume that would require mocking get_node_text. Let me know if you have any thoughts on this.

Reply:
You can have a look on the tests that were there when we removed `get_node_text?